src/models/geotimeline.py

```diff
@@ -1,6 +1,522 @@
+        """
+        Creates a timeline of multiple GeoGraphs, for time-series and change-detection
+        analyses.
+
+        The `data` must be a list of TimedGeoGraph objects or a dictionary where the
+        keys correspond to the respective time-stamps of each GeoGraph. The GeoGraphs
+        in `data` will be sorted from earliest to latest timestamp and added to the
+        GeoGraphTimeline.
+
+        Landscape, Habitat-component, Class-value and patch level time-series analyses
+        are supported.
+
+        Args:
+            data (Union[List[TimedGeoGraph], Dict[TimeStamp, GeoGraph]]): A list of
+                TimedGeoGraphs or a dictionary where keys correspond to times and values
+                to GeoGraph objects of the ecosystem at the specified time.
+
+        Raises:
+            NotImplementedError: For any other `data` argument. (In the future we will
+                add funcitonality to save timelines and load from disk.)
+        """
+
+        # Initialize empty graphs dict
+        self._graphs: Dict[TimeStamp, GeoGraph] = dict()
+        self.habitats: Dict[str, GeoGraphTimeline] = {}
+        # Fill the graphs dict with graphs from `data`
+        if isinstance(data, list):
+            self._load_from_sequence(graph_list=data)
+        elif isinstance(data, dict):
+            self._load_from_dict(graph_dict=data)
+        else:
+            raise NotImplementedError
+
+        # Initialize empty node map cache dictionary
+        self._node_map_cache: Dict[(TimeStamp, TimeStamp), NodeMap] = dict()
+
+    @property
+    def times(self) -> List[TimeStamp]:
+        """Return list of valid time stamps for this GeoGraphTimeline"""
+        return list(self._graphs.keys())
+
+    @property
+    def graphs(self) -> List[GeoGraph]:
+        """Return sorted list of GeoGraphs in this GeoGraphTimeline"""
+        return self._graphs
+
+    def __getitem__(self, time: TimeStamp) -> GeoGraph:
+        """Return the graph at this given time stamp in the GeoGraphTimeline"""
+        try:
+            return self._graphs[time]
+        except KeyError as error:
+            raise KeyError(f"Graph for time index `{time}` does not exist.") from error
+
+    def __len__(self) -> int:
+        """Return the number of timestamps in the GeoGraphTimeline"""
+        return len(self._graphs)
+
+    def __iter__(self) -> GeoGraph:
+        """Iterate over graphs in the GeoGraphTimeline in order (earlier to later)"""
+        return iter(self._graphs.values())
+
+    def _sort_by_time(self, reverse: bool = False) -> None:
+        """
+        Sort the graphs in GeoGraphTimeline accroding to timestamp from earlier to later
+
+        Args:
+            reverse (bool, optional): If False, start with the earliest date.
+                If True, sort starting with the latest date. Defaults to False.
+        """
+        self._graphs = {
+            time: self._graphs[time] for time in sorted(self._graphs, reverse=reverse)
```

Comment:
Nice to take advantage of the Python 3.7 feature to have ordered dicts by default!